account_fiscal_position_partner_domain/models/account_fiscal_position.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2021 ForgeFlow S.L.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from ast import literal_eval
+
+from odoo import api, fields, models
+
+
+class AccountFiscalPosition(models.Model):
+    _inherit = "account.fiscal.position"
+
+    partner_domain = fields.Char()
+
+    def _get_base_domain(self, vat_required, company_id, partner_domain=False):
+        base_domain = super()._get_base_domain(vat_required, company_id)
+        extra_domain = (
+            [("partner_domain", "!=", False)]
+            if partner_domain
+            else [("partner_domain", "=", False)]
+        )
+        return base_domain + extra_domain
+
+    @api.model
+    def _get_fpos_by_region_with_domain(
+        self,
+        country_id=False,
+        state_id=False,
+        zipcode=False,
+        vat_required=False,
+        partner_id=False,
+    ):
+        if not country_id:
+            return False
+        company_id = self.env.context.get("force_company", self.env.company.id)
+        base_domain = self._get_base_domain(vat_required, company_id, True)
+        null_state_dom = state_domain = [("state_ids", "=", False)]
+        null_zip_dom = zip_domain = [("zip_from", "=", False), ("zip_to", "=", False)]
+        null_country_dom = [
+            ("country_id", "=", False),
+            ("country_group_id", "=", False),
+        ]
+
+        if zipcode:
+            zip_domain = [("zip_from", "<=", zipcode), ("zip_to", ">=", zipcode)]
+
+        if state_id:
+            state_domain = [("state_ids", "=", state_id)]
+
+        domain_country = base_domain + [("country_id", "=", country_id)]
+        domain_group = base_domain + [("country_group_id.country_ids", "=", country_id)]
+
+        # Build domain to search records with exact matching criteria
+        fpos = self.search(domain_country + state_domain + zip_domain)
+        # return records that fit the most the criteria, and fallback on less specific
+        # fiscal positions if any can be found
+        if not fpos and state_id:
+            fpos = self.search(domain_country + null_state_dom + zip_domain)
+        if not fpos and zipcode:
+            fpos = self.search(domain_country + state_domain + null_zip_dom)
+        if not fpos and state_id and zipcode:
+            fpos = self.search(domain_country + null_state_dom + null_zip_dom)
+
+        # fallback: country group with no state/zip range
+        if not fpos:
+            fpos = self.search(domain_group + null_state_dom + null_zip_dom)
+
+        if not fpos:
+            # Fallback on catchall (no country, no group)
+            fpos = self.search(base_domain + null_country_dom)
+
+        # Evaluate domain in the Fiscal Positions that have been found
+        for fiscal_position in fpos:
+            partner_domain = literal_eval(fiscal_position.partner_domain)
```

Comment:
I think that here it would need to check if there is actually a partner_domain set in the fiscal position, as if it is empty ([]) it matches all records.

Reply:
I'm doing that here https://github.com/OCA/account-fiscal-rule/pull/212/files#diff-6cb6576d850e90ddac4f46fd15fcd258a9634a2cd4e06b8f21058f865ada2344R34

The base domain contains `[("partner_domain", "!=", False)]`, so all the Fiscal Positions found will have a partner_domain set.

Reply:
I see, the problem is that if you add some filters in the domain and then you remove them, the partner_domain is left with [], so that is when you have the issue.

Reply:
![image](https://user-images.githubusercontent.com/71635103/123921042-e40f8700-d986-11eb-9268-f8053dd8186c.png)
![image](https://user-images.githubusercontent.com/71635103/123921070-ebcf2b80-d986-11eb-827c-628f58f2429b.png)

Reply:
Nice catch :+1: 

I have updated the base_domain so it also filters out those Fiscal Position with partner_domain = [ ]